### PR TITLE
Bipolar sigmoid added

### PR DESCRIPTION
**Pull request by IWNMWE**
Implemented the bipolar sigmoid activation function
Fixed invertible activation functions 
Added jacobian testing for all of the above mentioned ones

Tested and used in various research papers:

https://www.scitepress.org/Papers/2018/100372/100372.pdf

https://iopscience.iop.org/article/10.1088/1742-6596/1255/1/012023/meta

### CODE REVIEW

**Review by zoq — COMMENTED** [review]
Let's add the same tests mentioned here - https://github.com/mlpack/mlpack/pull/3498#pullrequestreview-1486788244

**Comment by IWNMWE** [comment]
> 

Hey @zoq I had checked out the issue you had mentioned, I just wanted to ask about how we can tackle functions which do require both the x and f(x) value for the calculation of their derivatives ,a great example for this type of activation function is the EliSH function whose derivative can't be expressed solely in terms of f(x).(I have even checked other libraries and tried my self to solve for f'(x) in terms of f(x) but was not able to find any implementation that uses only f(x) values.)

A possible fix for these kinds of layers is to store the x values as a member of the class and use them in the deriv() method pls do let me know if there is any problem with this implementation or if there are any other fixes possible I will look into it.

**Comment by zoq** [comment]
Yes, you have ti implement it as a layer and store the activation function in the layer.

**Comment by IWNMWE** [comment]
@zoq Ok I will get to implementing all the required ones as layers. Then in this PR, I have fixed all activation functions which do have an inverse or can be brought into terms of their derivatives and add the jacobian test for all of them and in the next few PRs I will add the other ones as a layer

**Comment by IWNMWE** [comment]
The jacobian testing implemented is as same as given in #3478

**Comment by zoq** [comment]
> Hey @zoq can you please review the changes and check if the above mentioned plan is fine so that I can get to implementing the non invertible functions as layers

Looking at it right now.

File: src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/**
+ * @file methods/ann/activation_functions/bipolar_sigmoid_function.hpp
+ * @author Satyam Shukla
+ *
+ * Definition and implementation of the bipolar sigmoid  function.
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp:5]:
```suggestion
 * Definition and implementation of the bipolar sigmoid function.
```

File: src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ public:
+  /**
+   * Computes the Bipolar Sigmoid function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    return (1 - std::exp(-x))/(1 + std::exp(-x));
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp:39]:
```suggestion
    return (1 - std::exp(-x)) / (1 + std::exp(-x));
```

File: src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  /**
+   * Computes the bipolar sigmoid function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = (1 - arma::exp(-x))/(1 + arma::exp(-x));
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp:51]:
```suggestion
    y = (1 - arma::exp(-x)) / (1 + arma::exp(-x));
```

File: src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  /**
+   * Computes the first derivative of the Bipolar Sigmoid function.
+   *
+   * @param y Input activation.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    return (1.0 - std::pow(y,2)) / 2.0;
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp:62]:
```suggestion
    return (1.0 - std::pow(y, 2)) / 2.0;
```

File: src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  /**
+   * Computes the first derivatives of the Bipolar Sigmoid function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x)
+  {
+    x =  (1.0 - arma::pow(y,2)) / 2.0;
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/bipolar_sigmoid_function.hpp:74]:
```suggestion
    x =  (1.0 - arma::pow(y, 2)) / 2.0;
```

File: src/mlpack/methods/ann/activation_functions/hyper_sinh_function.hpp [hunk 6]
```diff
@@ -80,9 +80,9 @@ class HyperSinhFunction
   static double Deriv(const double y)
   {
     if (y > 0)
-      return (std::cosh(y) / 3);
+      return (std::pow((1.0 / 9.0) + (y * y),0.5));
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/hyper_sinh_function.hpp:83]:
```suggestion
      return (std::pow((1.0 / 9.0) + (y * y), 0.5));
```

File: src/mlpack/methods/ann/activation_functions/hyper_sinh_function.hpp [hunk 7]
```diff
@@ -95,15 +95,16 @@ class HyperSinhFunction
   static void Deriv(const InputVecType& y, OutputVecType& x)
   {
     x.set_size(y.size());
+    #pragma omp for
     for(size_t i = 0; i < y.n_elem; ++i)
     {
       if (y(i) > 0)
       {
-        x(i) = std::cosh(y(i)) / 3;
+        x(i) = (std::pow((1.0 / 9.0) + (y(i) * y(i)),0.5));
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/hyper_sinh_function.hpp:103]:
```suggestion
        x(i) = (std::pow((1.0 / 9.0) + (y(i) * y(i)), 0.5));
```

**Review by zoq — COMMENTED** [review]
Minor style issues, the test and implementation looks correct.

**Comment by zoq** [comment]
Let's rerun the CI.

**Comment by IWNMWE** [comment]
Hey can this be merged so that I can create the subsequent pull requests required

**Comment by zoq** [comment]
Sorry for the delay, thanks again for the contribution.

**Pull request merged**